File: src/Spreadsheet/ArcStudy.fs
namespace ARCtrl.Spreadsheet

open ARCtrl
open FsSpreadsheet

open ARCtrl.Helper

module ArcStudy = 

    let [<Literal>] obsoleteStudiesLabel = "STUDY METADATA"
    let [<Literal>] studiesLabel = "STUDY"

    let [<Literal>] obsoleteMetadataSheetName = "Study"
    let [<Literal>] metadataSheetName = "isa_study"

    let toMetadataSheet (study : ArcStudy) (assays : ArcAssay list option) : FsWorksheet =
        //let toRows (study:ArcStudy) assays =
        //    seq {          
        //        yield  SparseRow.fromValues [studiesLabel]
        //        yield! Studies.StudyInfo.toRows study
        //    }
        let sheet = FsWorksheet(metadataSheetName)
        Studies.toRows study assays
        |> Seq.append [SparseRow.fromValues [studiesLabel]]
        |> Seq.iteri (fun rowI r -> SparseRow.writeToSheet (rowI + 1) r sheet)    
        sheet

    let fromMetadataSheet (sheet : FsWorksheet) : ArcStudy*ArcAssay list =
        try
            let fromRows (rows: seq<SparseRow>) =
                let en = rows.GetEnumerator()
                en.MoveNext() |> ignore  
                let _,_,_,study = Studies.fromRows 2 en
                study
            sheet.Rows 
            |> Seq.map SparseRow.fromFsRow
            |> fromRows
            |> Option.defaultValue (ArcStudy.create(Identifier.createMissingIdentifier()),[])
        with 
        | err -> failwithf "Failed while parsing metadatasheet: %s" err.Message

    let toMetadataCollection (study : ArcStudy) (assays : ArcAssay list option) =
        Studies.toRows study assays
        |> Seq.append [SparseRow.fromValues [studiesLabel]]
        |> Seq.map (fun r -> SparseRow.getAllValues r)

    let fromMetadataCollection (collection: seq<seq<string option>>) : ArcStudy*ArcAssay list =
        try
            let fromRows (rows: seq<SparseRow>) =
                let en = rows.GetEnumerator()
                en.MoveNext() |> ignore  
                let _,_,_,study = Studies.fromRows 2 en
                study
            collection
            |> Seq.map SparseRow.fromAllValues
            |> fromRows
            |> Option.defaultValue (ArcStudy.create(Identifier.createMissingIdentifier()),[])
        with 
        | err -> failwithf "Failed while parsing metadatasheet: %s" err.Message

    let isMetadataSheetName (name : string) =
        name = metadataSheetName || name = obsoleteMetadataSheetName

    let isMetadataSheet (sheet : FsWorksheet) =
        isMetadataSheetName sheet.Name

    let tryGetMetadataSheet (doc:FsWorkbook) =
        doc.GetWorksheets()
        |> Seq.tryFind isMetadataSheet

[<AutoOpen>]
module ArcStudyExtensions =

    type ArcStudy with
    
        /// Reads an assay from a spreadsheet
        static member fromFsWorkbook (doc:FsWorkbook) = 
            try
                // Reading the "Assay" metadata sheet. Here metadata 
                let studyMetadata,assays =       
                    match ArcStudy.tryGetMetadataSheet doc with                     
                    | Option.Some sheet ->
                        ArcStudy.fromMetadataSheet sheet
                    | None -> 
                        printfn "Cannot retrieve metadata: Study file does not contain \"%s\" or \"%s\" sheet." ArcStudy.metadataSheetName ArcStudy.obsoleteMetadataSheetName
                        ArcStudy.create(Identifier.createMissingIdentifier()),[]
                let sheets = doc.GetWorksheets()
                let annotationTables = 
                    sheets
                    |> ResizeArray.choose ArcTable.tryFromFsWorksheet
                // Performance hotfix. This change is tested in ISA.Spreadsheet/Performance.Tests.fs and results in 2 pendings tests in ARCtrl/ARCtrl.Tests.fs.
                //if annotationTables |> Seq.isEmpty |> not then 
                //    let updatedTables = 
                //            ArcTables.updateReferenceTablesBySheets( // This only kills performance with ProtocolREF
                //                (ArcTables studyMetadata.Tables),
                //                (ArcTables (ResizeArray annotationTables)),
                //                keepUnusedRefTables =  true
                //                )
                //    studyMetadata.Tables <- updatedTables.Tables
                let datamapSheet =
                    sheets |> Seq.tryPick DataMapTable.tryFromFsWorksheet

                if annotationTables |> ResizeArray.isEmpty |> not then
                    studyMetadata.Tables <- annotationTables
                studyMetadata.DataMap <- datamapSheet

                studyMetadata,assays
            with
            | err -> failwithf "Could not parse study: \n%s" err.Message

        /// <summary>
        /// Write a study to a spreadsheet
        ///
        /// If datamapSheet is true, the datamap will be written to a worksheet inside study workbook. Default: true
        /// </summary>
        /// <param name="study"></param>
        /// <param name="assays"></param>
        /// <param name="datamapSheet"></param>
        static member toFsWorkbook (study : ArcStudy, ?assays : ArcAssay list, ?datamapSheet: bool) =
            let datamapSheet = defaultArg datamapSheet true
            let doc = new FsWorkbook()
            let metadataSheet = ArcStudy.toMetadataSheet study assays
            doc.AddWorksheet metadataSheet

            if datamapSheet then
                study.DataMap
                |> Option.iter (DataMapTable.toFsWorksheet >> doc.AddWorksheet)

            study.Tables
            |> Seq.iteri (fun i -> ArcTable.toFsWorksheet (Some i) >> doc.AddWorksheet)

            doc

        member this.ToFsWorkbook (?assays : ArcAssay list, ?datamapSheet: bool) =
            ArcStudy.toFsWorkbook (this, ?assays = assays, ?datamapSheet = datamapSheet)

File: js/Spreadsheet/ArcStudy.js
import { FsWorksheet } from "../fable_modules/FsSpreadsheet.6.3.0-alpha.4/FsWorksheet.fs.js";
import { iterate, tryPick, tryFind, map, append, iterateIndexed } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { SparseRowModule_fromAllValues, SparseRowModule_getAllValues, SparseRowModule_fromFsRow, SparseRowModule_fromValues, SparseRowModule_writeToSheet } from "./Metadata/SparseTable.js";
import { fromRows, toRows } from "./Metadata/Study.js";
import { getEnumerator } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { toConsole, printf, toFail } from "../fable_modules/fable-library-js.4.22.0/String.js";
import { unwrap, toArray, defaultArg } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { ArcStudy } from "../Core/ArcTypes.js";
import { createMissingIdentifier } from "../Core/Helper/Identifier.js";
import { empty } from "../fable_modules/fable-library-js.4.22.0/List.js";
import { ResizeArray_isEmpty, ResizeArray_choose } from "../Core/Helper/Collections.js";
import { toFsWorksheet as toFsWorksheet_1, tryFromFsWorksheet } from "./AnnotationTable/ArcTable.js";
import { toFsWorksheet, tryFromFsWorksheet as tryFromFsWorksheet_1 } from "./DataMapTable/DataMapTable.js";
import { FsWorkbook } from "../fable_modules/FsSpreadsheet.6.3.0-alpha.4/FsWorkbook.fs.js";

export function ArcStudy_toMetadataSheet(study, assays) {
    let source_1;
    const sheet = new FsWorksheet("isa_study");
    iterateIndexed((rowI, r) => {
        SparseRowModule_writeToSheet(rowI + 1, r, sheet);
    }, (source_1 = toRows(study, assays), append([SparseRowModule_fromValues(["STUDY"])], source_1)));
    return sheet;
}

export function ArcStudy_fromRows(rows) {
    const en = getEnumerator(rows);
    en["System.Collections.IEnumerator.MoveNext"]();
    return fromRows(2, en)[3];
}

export function ArcStudy_fromMetadataSheet(sheet) {
    try {
        return defaultArg(ArcStudy_fromRows(map(SparseRowModule_fromFsRow, sheet.Rows)), [ArcStudy.create(createMissingIdentifier()), empty()]);
    }
    catch (err) {
        const arg = err.message;
        return toFail(printf("Failed while parsing metadatasheet: %s"))(arg);
    }
}

export function ArcStudy_toMetadataCollection(study, assays) {
    let source_1;
    return map(SparseRowModule_getAllValues, (source_1 = toRows(study, assays), append([SparseRowModule_fromValues(["STUDY"])], source_1)));
}

export function ArcStudy_fromMetadataCollection(collection) {
    try {
        return defaultArg(ArcStudy_fromRows(map(SparseRowModule_fromAllValues, collection)), [ArcStudy.create(createMissingIdentifier()), empty()]);
    }
    catch (err) {
        const arg = err.message;
        return toFail(printf("Failed while parsing metadatasheet: %s"))(arg);
    }
}

export function ArcStudy_isMetadataSheetName(name) {
    if (name === "isa_study") {
        return true;
    }
    else {
        return name === "Study";
    }
}

export function ArcStudy_isMetadataSheet(sheet) {
    return ArcStudy_isMetadataSheetName(sheet.Name);
}

export function ArcStudy_tryGetMetadataSheet(doc) {
    return tryFind(ArcStudy_isMetadataSheet, doc.GetWorksheets());
}

/**
 * Reads an assay from a spreadsheet
 */
export function ARCtrl_ArcStudy__ArcStudy_fromFsWorkbook_Static_32154C9D(doc) {
    try {
        let patternInput;
        const matchValue = ArcStudy_tryGetMetadataSheet(doc);
        if (matchValue == null) {
            toConsole(printf("Cannot retrieve metadata: Study file does not contain \"%s\" or \"%s\" sheet."))("isa_study")("Study");
            patternInput = [ArcStudy.create(createMissingIdentifier()), empty()];
        }
        else {
            patternInput = ArcStudy_fromMetadataSheet(matchValue);
        }
        const studyMetadata = patternInput[0];
        const sheets = doc.GetWorksheets();
        const annotationTables = ResizeArray_choose(tryFromFsWorksheet, sheets);
        const datamapSheet = tryPick(tryFromFsWorksheet_1, sheets);
        if (!ResizeArray_isEmpty(annotationTables)) {
            studyMetadata.Tables = annotationTables;
        }
        studyMetadata.DataMap = datamapSheet;
        return [studyMetadata, patternInput[1]];
    }
    catch (err) {
        const arg_2 = err.message;
        return toFail(printf("Could not parse study: \n%s"))(arg_2);
    }
}

/**
 * Write a study to a spreadsheet
 * 
 * If datamapSheet is true, the datamap will be written to a worksheet inside study workbook. Default: true
 */
export function ARCtrl_ArcStudy__ArcStudy_toFsWorkbook_Static_Z4CEFA522(study, assays, datamapSheet) {
    const datamapSheet_1 = defaultArg(datamapSheet, true);
    const doc = new FsWorkbook();
    const metadataSheet = ArcStudy_toMetadataSheet(study, assays);
    doc.AddWorksheet(metadataSheet);
    if (datamapSheet_1) {
        iterate((arg) => {
            const sheet = toFsWorksheet(arg);
            doc.AddWorksheet(sheet);
        }, toArray(study.DataMap));
    }
    iterateIndexed((i, arg_1) => {
        const sheet_1 = toFsWorksheet_1(i, arg_1);
        doc.AddWorksheet(sheet_1);
    }, study.Tables);
    return doc;
}

export function ARCtrl_ArcStudy__ArcStudy_ToFsWorkbook_257FC1F0(this$, assays, datamapSheet) {
    return ARCtrl_ArcStudy__ArcStudy_toFsWorkbook_Static_Z4CEFA522(this$, unwrap(assays), unwrap(datamapSheet));
}

